semcore/d3-chart/src/component/Chart/BubbleChart.tsx

```diff
@@ -90,7 +69,37 @@ class BubbleChartComponent extends AbstractChart<BubbleChartData, BubbleChartPro
 
     const range = [marginY, plotWidth - this.plotPadding];
 
-    return scaleLinear([Math.min(...this.scaledValues.x), Math.max(...this.scaledValues.x)], range);
+    const miniestValue = data.reduce(
+      (acc, item) => {
+        if (item.x - item.value < acc.xMin) {
+          acc.xMin = item.x - item.value;
+          acc.value = item.value;
+        }
+        return acc;
+      },
+      { value: data[0].value, xMin: data[0].x - data[0].value },
+    ).value;
+    const maxestValue = data.reduce(
+      (acc, item) => {
+        if (item.value + item.x > acc.xMax) {
+          acc.xMax = item.value + item.x;
+          acc.value = item.value;
+        }
+        return acc;
+      },
+      { value: data[0].value, xMax: data[0].value + data[0].x },
+    ).value;
+
+    let xMin = Math.min(...data.map((d) => d.x));
+    let xMax = Math.max(...data.map((d) => d.x));
+    const domainWidh = Math.abs(xMax - xMin);
+    const pixelRactio = domainWidh / Math.abs(range[0] - range[1]);
+    const minShift = this.valueScale(miniestValue) * pixelRactio;
+    const maxShift = this.valueScale(maxestValue) * pixelRactio;
+    xMin -= minShift * 2;
+    xMax += maxShift * 2;
+
+    return scaleLinear().domain([xMin, xMax]).range(range);
```

Comment:
let's add function for calculate domain or all scale? 
`calcScale(data: BubbleChartData, key: 'x' | 'y', range: [number, number]): ScaleLinear`

Reply:
Refactored